### PR TITLE
Add new `attribute_name_spacing` rule

### PR DESCRIPTION
**Pull request by aryamansharda**
Resolves https://github.com/realm/SwiftLint/issues/5667

Seeing as `private (set)` would cause issues with the Swift 6 compiler in Xcode 16+, this PR implements the changes discussed in the issue linked above. 

The rule is now a default rule and prevents the following code from triggering errors:

Triggers:
`private (set) var type: ClipType = .undetermined`

Doesn't Trigger:
`private(set) var type: ClipType = .undetermined`

First time contributing to SwiftLint, so apologies for any contribution mistakes on my end. You managed to convince me to try and contribute @SimplyDanny (re. SwiftCraft)

### CODE REVIEW

**Comment by SimplyDanny** [comment]
Hey @aryamansharda, nice to hear from you and even better via a contribution!

The implementation looks good to me. Yet, checking the commits that introduced this warning/error in the [Swift compiler](https://github.com/swiftlang/swift/commit/c6e425a55938dba840416e2ef1ae1491fc048082) and SwiftSyntax, I just found out that `<access-control-modifier> (set)` is only one special case. The warning/error is actually more general and applies to all modifiers and attributes likewise, e.g. `@available (*, deprecated)` or `nonisolated (unsafe)`.

For the attribute case, we could also check that there is no trivia between the `@` sign and the attribute name as this is another related compiler diagnostic. It also seems that while
```swift
let closure1 = { @MainActor (a, b) in
}
```
leads to the warning/error,
```swift
let closure2 = { @MainActor 
   (a, b) in
}
```
does not. In other words, a line break serves as a disambiguator.

I hope this isn't too much to ask for, but it seems that only supporting all these cases makes this rule a real help in preparation for Swift 6. I'm happy to help in case of any issues. 😉

**Comment by aryamansharda** [comment]
Sounds good! I'll explore that today - just wanted to double-check my understanding before I continue. 

First, I should should update my existing implementation to check that there is no trivia between any `DeclModifier` and `DeclModifierDetail` which would handle cases like `(set)` and `nonisolated(unsafe)`

Next, based off the examples in the commit you referenced:
```
struct PropertyWrapperTest {
  @MyPropertyWrapper (param: 2)  // expected-warning {{extraneous whitespace between attribute name and '('; this is an error in Swift 6}}
  var x: Int
}

let closure1 = { @MainActor (a, b) in // expected-warning {{extraneous whitespace between attribute name and '('; this is an error in Swift 6}}
}

let closure2 = { @MainActor
  (a: Int, b: Int) in
}
```
When I plug them into the Swift AST Explorer, I see that `AttributeList` and `Attribute` are the type we're after. 

So, in addition to the existing check above, we want to add a new check for all `AttributeList->Attribute->IdentifierType` to ensure that there is no leading (ex. `@ MainActor` / `@ escaping`) or trailing trivia (ex. `@MyPropertyWrapper (param: 2)` / ` @MainActor (a, b)`)?

**Comment by SimplyDanny** [comment]
Yes, you mentioned all the cases we can find in the related Swift compiler diagnostics.

**Comment by mildm8nnered** [comment]
Not sure if this will work or not, but you might be able to avoid the Rewriter implementation by doing something like [ImplicitReturnRule](https://github.com/realm/SwiftLint/blob/cc4b569c5491b33963f3e0f435e7722f9311b77e/Source/SwiftLintBuiltInRules/Rules/Style/ImplicitReturnRule.swift#L1) does with `SwiftSyntaxCorrectableRule`.

**Comment by aryamansharda** [comment]
@mildm8nnered Ah, very cool. Just updated the implementation to use that 👍

File: Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift:36]: Let's check if `@MyProperty ()` triggers as well. 🤓

**Review by SimplyDanny — CHANGES_REQUESTED** [review]
Looks good to me! All the main cases are covered. I have only a few small remarks.

**Comment by SimplyDanny** [comment]
Hey @aryamansharda, anything that's blocking this PR? I'd like to include it into the next major release. Swift 6 is not too far off. To drive this rule useable, we shouldn't wait for too long.

**Comment by aryamansharda** [comment]
Hey, sorry for the delay! I'll make the updates in the next day or two!

**Comment by aryamansharda** [comment]
@SimplyDanny Sorry for the delay, but I've made the requested changes. If there's any other changes or documentation changes let me know - I promise the following turn around time will be much faster.

File: CHANGELOG.md
Comment (SimplyDanny) [inline CHANGELOG.md:132]: Please append two spaces to the end of the description. They are required to enforce a line break in the rendered Markdown.

```suggestion
* Add new `access_control_setter_spacing` rule to enforce no trailing whitespace between
  attribute names and parentheses, ensuring compatibility with Swift 6, where this spacing
  causes compilation errors.  
  [aryamansharda](https://github.com/aryamansharda)
  [#5667](https://github.com/realm/SwiftLint/issues/5667)

```

File: Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift:11]: Use a string block:
```suggestion
        description: """
            This rule prevents trailing spaces after attribute names, ensuring compatibility \
            with Swift 6 where a space between an attribute name and the opening parenthesis \
            results in a compilation error (e.g. `@MyPropertyWrapper ()`, `private (set)`).
            """,
```

File: Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift:48]:
```suggestion
            Example("@MyProperty ↓() class Foo {}"),
```

File: Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift:135]: The description of this rule is very detailed which is good. It's impractical as a violation message, though. Better to set an explicit reason here that depends on the actual violation and is much shorter. That would look like:
```swift
violations.append(
    .init(
        position: endPosition, 
        reason: reason, 
        severity: configuration.severity, 
        correction: .init(
            start: startPosition,
            end: endPosition,
            replacement: replacement
        )
    )
)
```

Incidentally, this is also why the tests fail.

**Review by SimplyDanny — CHANGES_REQUESTED** [review]

File: Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift
Comment (aryamansharda) [reply Source/SwiftLintBuiltInRules/Rules/Style/AttributeNameSpacingRule.swift]: Updated!

**Review by SimplyDanny — APPROVED** [review]

**Comment by SimplyDanny** [comment]
Thanks @aryamansharda! I only made the wording more serious to better match the `error` severity and the fact that these violation will be compiler errors in the future.

Once the build passes, this is ready to be merged.

**Pull request merged**
